### PR TITLE
#63 add statistics

### PR DESCRIPTION
**Pull request by lojzatran**
Fixes: https://github.com/commercetools/commercetools-node-variant-reassignment/issues/63

### CODE REVIEW

File: lib/runner/variant-reassignment.js [hunk 1]
```diff
@@ -86,6 +98,7 @@ export default class VariantReassignment {
         { concurrency: 3 })
       return this.errorCallback(error, productDraft)
     }
+    this.statistics.retries++
     const transactions = await this.transactionService.getTransactions()
     const failedTransaction = transactions.find(({ value }) =>
         _.isEqual(value.newProductDraft.name, productDraft.name)
```
Comment (butenkor) [inline lib/runner/variant-reassignment.js]: Why match transactions by product name where product name is not unique? Why not slug for example?

Comment (butenkor) [inline lib/runner/variant-reassignment.js]: Can here be an else case?

**Comment by butenkor** [comment]
@lojzatran Why did you merge without any approval?

**Comment by lojzatran** [comment]
It's merged only to a testing branch, not master. I wanted to test it in the sphere-product-import.

**Comment by butenkor** [comment]
@lojzatran When and where  it is supposed to be reviewed then?

**Comment by lojzatran** [comment]
It should be reviewed here and now. I didn't want to make a base master branch, because then it would mean too many changes. I also didn't know when you will review this, so I merge it to the testing branch, so I can test it with product-import in the meantime.

File: lib/runner/variant-reassignment.js [hunk 2]
```diff
@@ -25,7 +33,7 @@ export default class VariantReassignment {
    *  - if yes, create and process actions which will move variants across products
    * @param productDrafts List of productDrafts
    * @param productTypeNameToTypeObj product type cache to resolve product type references
-   * @returns {Promise.<*>} true if reassignment has been executed
+   * @returns {Promise.<*>} total reassignment statistics of all runs
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: It means that if you run this method multiple times, the statistics won't reset. To reset stats, you need to create a new instance of variant reassignment.

File: lib/runner/variant-reassignment.js (hunk 1 above)
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: What do you mean by `else` case?

File: lib/runner/variant-reassignment.js [hunk 3]
```diff
@@ -34,6 +42,7 @@ export default class VariantReassignment {
       if (this.firstRun)
         await this._processUnfinishedTransactions()
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: https://github.com/commercetools/commercetools-node-variant-reassignment/commit/8d6347df83db648213ebf1a3f646d19ba318b2fb

File: lib/runner/variant-reassignment.js [hunk 4]
```diff
@@ -34,6 +42,7 @@ export default class VariantReassignment {
...
     } catch (e) {
+      this.statistics.errors++
       return this._error('Could not process unfinished transactions', e)
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: https://github.com/commercetools/commercetools-node-variant-reassignment/commit/4f65fe2802c7366ab91d964eadafd820a94b1d66

File: lib/runner/variant-reassignment.js [hunk 5]
```diff
@@ -78,6 +89,7 @@ export default class VariantReassignment {
    * @see https://github.com/commercetools/commercetools-node-variant-reassignment/issues/60
    */
   async _handleProcessingError (productDraft, products, error) {
+    this.statistics.errors++
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: https://github.com/commercetools/commercetools-node-variant-reassignment/commit/4c8827f7d9014c2ae47aa9856c3f35564c259fce#diff-eda63119a30637151d12307cefd81cf0

File: lib/runner/variant-reassignment.js (hunk 1 above)
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: https://github.com/commercetools/commercetools-node-variant-reassignment/commit/e72a6f5cf4350a509ab9710fae9f284d1981d680

File: lib/runner/variant-reassignment.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (butenkor) [reply lib/runner/variant-reassignment.js]: In linked commit i do not find any 409 handling but i commented there on other topics.

File: lib/runner/variant-reassignment.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (butenkor) [reply lib/runner/variant-reassignment.js]: `total reassignment statistics of all runs` -> `Updated global statistics`

File: lib/runner/variant-reassignment.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (butenkor) [reply lib/runner/variant-reassignment.js]: Commented on linked commit

File: lib/runner/variant-reassignment.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: https://github.com/commercetools/commercetools-node-variant-reassignment/commit/54745e53b1f8e8f9f5884d1955330e5278bfb99a

**Comment by lojzatran** [comment]
@butenkor I changed the repeater method: https://github.com/commercetools/commercetools-node-variant-reassignment/commit/b5acc2378230153fdc39c83200d7e17cd34add2d

**Pull request merged**